refactor(migrations): support use of an ESM `@angular/compiler-cli` package

clydin opened the pull request:
Currently, migrations and schematics must be in CommonJS format. However, framework packages will only be ESM from v13 and onward. To support this configuration, dynamic import expressions are now used to load `@angular/compiler-cli` and its new secondary entry point `@angular/compiler-cli/private/migrations`. Dynamic imports within Node.js allow the `@angular/core` migrations’ CommonJS code to load ESM code. Unfortunately, TypeScript will currently, unconditionally down-level dynamic import into a require call. `require` calls cannot load ESM code and will result in a runtime error. To workaround this, a Function constructor is used to prevent TypeScript from changing the dynamic import. Once TypeScript provides support for keeping the dynamic import this workaround can be dropped and replaced with a standard dynamic import.  Due to the use of the dynamic import, a reference to the dynamically loaded modules must now be passed to all locations that use values from the modules.

clydin committed:
refactor(migrations): support use of an ESM `@angular/compiler-cli` package

Currently, migrations and schematics must be in CommonJS format. However, framework packages will only be ESM from v13 and onward. To support this configuration, dynamic import expressions are now used to load `@angular/compiler-cli` and its new secondary entry point `@angular/compiler-cli/private/migrations`. Dynamic imports within Node.js allow the `@angular/core` migrations’ CommonJS code to load ESM code. Unfortunately, TypeScript will currently, unconditionally down-level dynamic import into a require call. `require` calls cannot load ESM code and will result in a runtime error. To workaround this, a Function constructor is used to prevent TypeScript from changing the dynamic import. Once TypeScript provides support for keeping the dynamic import this workaround can be dropped and replaced with a standard dynamic import.  Due to the use of the dynamic import, a reference to the dynamically loaded modules must now be passed to all locations that use values from the modules.

devversion reviewed (approved):
LGTM. Hopefully we can use ESM for the migrations in the future as this would simplify things a lot.

clydin committed:
fixup! refactor(migrations): support use of an ESM `@angular/compiler-cli` package

clydin committed:
refactor(migrations): ensure CommonJS migrations can be accessed

The `@angular/core` migrations are currently published as CommonJS and also not bundled. Schematics, of which migrations are a type, are currently required to be CommonJS modules due to the Schematics Runtime not yet supporting ESM-based schematics. To ensure that the migration files are loaded as CommonJS modules, a nested `package.json` file has been introduced within the `schematics` directory of the package to set the `type` field for all containing JavaScript files as CommonJS. Since the migrations are also not currently bundled and the `devmode` output used to build the migrations will replace all relative imports with fully-qualified module specifiers, an additional `exports` entry must be added to allow the fully-qualified module specifiers to correctly resolve. If `devmode` did not change the relative imports the additional entry would not be needed. Bundling would also remedy the situation and is the long-term path, especially once the migrations are converted to ESM. Additional followup changes should investigate bundling each migration. The additional `exports` entry should be removed if either `devmode` behavior is changed or the migrations are bundled.

devversion commented on packages/core/schematics/utils/load_esm.ts:
```diff
@@ -69,15 +41,15 @@ export async function loadEsmModule<T>(modulePath: string|URL): Promise<T> {
+    // TODO_ESM: This can be removed once Bazel supports exports fields.
     return {
       ...require('@angular/compiler-cli/src/ngtsc/annotations'),
       ...require('@angular/compiler-cli/src/ngtsc/imports'),
```
I think the fallback should could just be: `require('@angular/compiler-cli/src/ngtsc/migrations.js);` (notice the explicit extension)

devversion reviewed (approved):
Still LGTM!

clydin committed:
fixup! refactor(migrations): support use of an ESM `@angular/compiler-cli` package

clydin committed:
fixup! refactor(migrations): ensure CommonJS migrations can be accessed

clydin committed:
fixup! refactor(migrations): ensure CommonJS migrations can be accessed

alxhub commented:
Caretaker: cl/400826396 is a necessary update to the g3 patch for this change. Please coordinate with me to land these together.

dylhunn commented:
This PR was merged into the repository by commit 6d1c2f79da9e034308c7f2cfbd16499e0cfd51f9.
